### PR TITLE
Simplify constant handling in Painless

### PR DESCRIPTION
**Pull request by jdconrad**
This change makes constant handling simpler by doing the following:
1) Removes the ir tree node RegexNode entirely in favor or using more generic nodes such as ConstantNode and MemberFieldLoadNode generated by the user tree node ERegex. This requires a clinit method added to the ClassNode, but allows for additional removals. This clinit method can also be used for potential optimization passes.
2) Removes the Constant class as with the changes to regex nodes it's no longer necessary.
3) Removes the Globals class from the ir tree write phase. With the previous removals, Globals is no longer necessary.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java [hunk 1]
```diff
@@ -72,6 +72,7 @@ StatementExpressionNode write(ClassNode classNode) {
 
         statementExpressionNode.setLocation(location);
         statementExpressionNode.setMethodEscape(methodEscape);
+        statementExpressionNode.setNoop(false);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java]: Why is this necessary?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/MemberFieldLoadNode.java [hunk 2]
```diff
@@ -51,12 +50,8 @@ public boolean isStatic() {
 
     /* ---- end node data ---- */
 
-    public MemberFieldNode() {
-        // do nothing
-    }
-
     @Override
-    public void write(ClassWriter classWriter, MethodWriter methodWriter, Globals globals, ScopeTable scopeTable) {
+    public void write(ClassWriter classWriter, MethodWriter methodWriter, ScopeTable scopeTable) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/MemberFieldLoadNode.java]: There's now `MemberFieldLoadNode` and `MemberFieldStoreNode` where `Store` has the following new code:
```java
        if (isStatic == false) {
            methodWriter.loadThis();
        }

        getChildNode().write(classWriter, methodWriter, scopeTable);
```

Why the new code?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java [hunk 3]
```diff
@@ -74,30 +81,90 @@ void analyze(ScriptRoot scriptRoot, Scope scope) {
...
+            staticNode.setExpressionType(Pattern.class);
+
+            callNode.setLeftNode(staticNode);
+
+            CallSubNode callSubNode = new CallSubNode();
+            callSubNode.setLocation(location);
+            callSubNode.setExpressionType(Pattern.class);
+            callSubNode.setBox(Pattern.class);
+            callSubNode.setMethod(new PainlessMethod(
+                    Pattern.class.getMethod("compile", String.class, int.class),
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java]: Where is this being written to `clinit`?

Comment (stu-elastic) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java]: Oh perhaps in `classNode.addFieldNode(fieldNode);`?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java]: After speaking with you (@stu-elastic) and @rjernst I've decided to change this a bit. I've removed `noop` in favor of `doPop` which will pop extraneous values off the stack when necessary such as a method call with side effects, but an ignored return value `somecall();`.

I also removed the need for StatementExpressionNode to handle return values. Instead, SExpression will just directly translate to a ReturnNode if a return is required. This moves the responsibility to where is makes sense instead of having redundant returns in multiple nodes.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/MemberFieldLoadNode.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/MemberFieldLoadNode.java]: The new code is to handle putting the value on the stack that will be written to the field. So store requires an expression to generate a value to store where as load does not.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java]: `classNode.addFieldNode` creates a declaration for a member field.

These lines actually add the regex constant generation and assignment to the clinit method:

```
BlockNode blockNode = classNode.getClinitNode().getBlockNode();
blockNode.addStatementNode(statementExpressionNode);
```

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/MemberFieldStoreNode.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+package org.elasticsearch.painless.ir;
+
+import org.elasticsearch.painless.ClassWriter;
+import org.elasticsearch.painless.MethodWriter;
+import org.elasticsearch.painless.symbol.ScopeTable;
+
+import static org.elasticsearch.painless.WriterConstants.CLASS_TYPE;
+
+public class MemberFieldStoreNode extends UnaryNode {
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/MemberFieldStoreNode.java]: Added.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/SExpression.java]: I continued to iterate on this after more discussion with @rjernst . I removed the doPop and will instead use an expression's type to determine if a pop is required or not (popping the void type is the same as a noop in this case).

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ClassNode.java [hunk 5]
```diff
@@ -152,34 +169,25 @@ public ScriptRoot getScriptRoot() {
         constructor.returnValue();
         constructor.endMethod();
 
+        if (clinitNode.getBlockNode().getStatementsNodes().isEmpty() == false) {
+            ReturnNode returnNode = new ReturnNode();
+            returnNode.setLocation(new Location("internal$clinit$return", 0));
+            clinitNode.getBlockNode().addStatementNode(returnNode);
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ClassNode.java]: I changed this to do no modifications of the ir tree from write. I just do the method writer for clinit and the return statement in the ClassNode. I'll consider better ways to do this, but for now this makes the most sense because it's the simplest change.

**Comment by jdconrad** [comment]
@rjernst @stu-elastic Thanks for the initial rounds of review.

**Comment by jdconrad** [comment]
@rjernst I think this is ready for another look. I believe I addressed all your comments and continued iterating on the best way to handle StatementExpressionNode - see Stu's initial comments on this for my follow ups.

**Review by rjernst — APPROVED** [review]
LGTM

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by jdconrad** [comment]
@rjernst Thanks for the second review.
